Simplify handling of signed ctrl messages.

kormat opened the pull request:
This patch reworks how scion ctrl payloads are created/signed/verified.
By using callbacks, an app can create and sign a ctrl payload and only
have a single error value to check, greatly reducing boiler-plate. It
also becomes much more obvious when a message is not being signed.

This patch also adds support for non-union data at the CtrlPld, CertMgmt
and PathMgmt levels. This allows for a unified approach to things like
request IDs, tracing, and error reporting.

Also:
- Cache the parsed CtrlPld in SignedCtrlPld, to reduce the overhead of
  extracting it multiple times.
- Add request ID and trace ID to CtrlPld, though they are currently not
  used.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/scionproto/scion/1385)
<!-- Reviewable:end -->

shitz commented:
Reviewed 16 of 16 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[go/lib/ctrl/ctrl.go, line 47 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46UeHVrrU2xGtOr92A:-L46UeHVrrU2xGtOr92B:b-p0dr9e) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl.go#L47)):*
> ```Go
> 
> // NewPldF returns a Plder that calls NewPld
> func NewPldF(u proto.Cerealizable, d *Data) Plder {
> ```

I'm not a fan of this pattern. The only thing you gain is that the caller doesn't have to first create the payload, check for an error, then call `Request` and check for an error and instead can just check the error of `Request`. However,  it took me a while to understand this. 

---

*[go/lib/ctrl/signed_util.go, line 24 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46cKNxpBXBnrbM1tsA:-L46cKNxpBXBnrbM1tsB:b2u3wqe) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L24)):*
> ```Go
> )
> 
> type Signer func(*Pld) (*SignedPld, error)
> ```

I wonder if all of these callbacks and closures should instead be `interfaces`. I learned that you can implement an interface with purely a method so this would allow your approach, but would also allow an object-based implementation plus I think interfaces are more intuitive to understand than closures.

---

*[go/lib/ctrl/signed_util.go, line 36 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46ZFwREQAVTWea447E:-L46ZFwREQAVTWea447F:blbu460) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L36)):*
> ```Go
> }
> 
> type trustStore struct{} // TODO(kormat): replace this with the actual trust store.
> ```

I _think_ this will lead to a circular dependency. Having the trust store here means importing `go/lib/trust`. The trust store itself will make use of `go/lib/ctrl`, either directly or through the messenger, which will then lead to a circular import. I'm not sure if these utils can be in `ctrl`

---

*[go/lib/ctrl/signed_util.go, line 41 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46Y9sqGyunQWaOPPon:-L46Y9sr-VWQ63z7XtE2:b-6mrmnx) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L41)):*
> ```Go
> type SigCheckerInner func(*SignedPld, *trustStore) error
> 
> func MkSigchecker(sigCin SigCheckerInner, tStore *trustStore) SigChecker {
> ```

`MkSigChecker`

---

*[go/lib/ctrl/signed_util.go, line 94 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46_ct33uPO-Mklx2QI:-L46_ct33uPO-Mklx2QJ:bdorwpo) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L94)):*
> ```Go
> 
> // MkSPld creates a SignedPld from the provided Plder and Signer.
> func MkSPld(plder Plder, signer Signer) (*SignedPld, error) {
> ```

This is a bit too acronym-y for my tasted. `MkSignedPld`is reasonably short. Also, given how often I see `ctrl.MkSPld(..., ctrl.NullSigner)` it probably would make sense to have something like `MkUnsignedPld`, even though it might be confusing that it returns an object of `SignedPld`. I guess the correct term would be `MkUnsignedSignedPld`, and the correct name for `SignedPld` would be `PossiblySignedPld` ;)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1385)*
<!-- Sent from Reviewable.io -->

scrye commented:
Reviewed 1 of 16 files at r1.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[go/lib/ctrl/ctrl.go, line 47 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46Kt6B4Az_jQEs6myW:-L46Kt6B4Az_jQEs6myX:b-dzzw9h) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl.go#L47)):*
> ```Go
> 
> // NewPldF returns a Plder that calls NewPld
> func NewPldF(u proto.Cerealizable, d *Data) Plder {
> ```

The state of the closure returned by this function is immutable and bound to variables `u` and `d`. This means that prior to each call to `Request`, the callback needs to be constructed to bind the variables:

```
plder = NewPldF(u, d)
ctrl_msg.Request(ctx, plder, ...)
```

At that point, why not just:
```
pld, err = NewPld(u, d)
// Handle err
ctrl_msg.Request(ctx, pld, ...)
```

The closure callback just moves error handling from the outside to inside `Request`, but it looks like it does nothing else.

More complex implementations of `Plder` are also subject to the same limitation, because their whole state needs to be constructed prior to each call to `Request`.

---

*[go/lib/ctrl/ctrl.go, line 96 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46IZHbPrnFQWeUz0SP:-L46IZHcNIdKQggXQ94v:b-l5uz5s) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl.go#L96)):*
> ```Go
> }
> 
> func (p *Pld) GetCertMgmt() (*cert_mgmt.Pld, *Data, error) {
> ```

The returned values here probably need a docstring.

---

*[go/lib/ctrl/ctrl.go, line 109 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46Ic75uhSZ362SPfBN:-L46Ic75uhSZ362SPfBO:b-l5uz5s) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl.go#L109)):*
> ```Go
> }
> 
> func (p *Pld) GetPathMgmt() (*path_mgmt.Pld, *Data, error) {
> ```

The returned values here probably need a docstring.

---

*[go/lib/ctrl/signed_util.go, line 38 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46d5-XNrQMYisbuVl5:-L46d5-XNrQMYisbuVl6:bkwhaf4) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L38)):*
> ```Go
> type trustStore struct{} // TODO(kormat): replace this with the actual trust store.
> 
> type SigChecker func(*SignedPld) error
> ```

Suggest changing `SigChecker` to `Verifier`, so it's more intuitive for external packages to refer to `ctrl.Signer` and `ctrl.Verifier`.

---

*[go/lib/ctrl/signed_util.go, line 39 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46VjmPZIKOB3x-cLdZ:-L46VjmPZIKOB3x-cLd_:bg7qw0b) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L39)):*
> ```Go
> 
> type SigChecker func(*SignedPld) error
> type SigCheckerInner func(*SignedPld, *trustStore) error
> ```

I'm not sure I understand what the meaning behind this type is. To me it looks like it is a variant of `SigChecker` with a bound trust store, but it is used after some checks in `MkSigchecker`. Where does the job of `SigChecker` internals stop, and the job of `SigCheckerInner` start?

---

*[go/lib/ctrl/signed_util.go, line 41 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46ScDoga6GYcq8adXe:-L46ScDpp0agUJsonL0y:b-bvl0u2) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L41)):*
> ```Go
> type SigCheckerInner func(*SignedPld, *trustStore) error
> 
> func MkSigchecker(sigCin SigCheckerInner, tStore *trustStore) SigChecker {
> ```

The purpose of `sigCin` is not evident to me. A very detailed docstring is needed here, s.t. users of package `ctrl` have an intuition on the differences between `SigChecker` and `SigCheckerInner`.

---

*[go/lib/ctrl/ctrl_msg/msg.go, line 15 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46eDH0041pSOqiXHPq:-L46eDH0041pSOqiXHPr:bviwckr) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl_msg/msg.go#L15)):*
> ```Go
> // limitations under the License.
> 
> package ctrl_msg
> ```

This needs a docstring, as it's not immediately clear what the package offers (it's more related to signing and verifying, than actually messages).

---

*[go/lib/ctrl/ctrl_msg/msg.go, line 78 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46ffeeiplQEa1sBu1I:-L46ffeeiplQEa1sBu1J:b-g166ew) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl_msg/msg.go#L78)):*
> ```Go
> }
> 
> func (r *Requester) notify(ctx context.Context, plder ctrl.Plder, a net.Addr, f notifyF) error {
> ```

I think the callback pattern is a bit more confusing than just calling a helper method to set up `spld` at the start of both `Notify` and `NotifyUnreliable`, and then calling the correct function. The reason being that to understand`Notify` and `NotifyUnreliable` a programmer needs to look into `notify`, and the other way around. In any case, I guess it's about preference here, and I prefer only using callbacks only when there are no other solutions.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1385)*
<!-- Sent from Reviewable.io -->

kormat commented:
Review status: 8 of 16 files reviewed at latest revision, 13 unresolved discussions.

---

*[go/lib/ctrl/ctrl.go, line 47 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46Kt6B4Az_jQEs6myW:-L47-ImywwLGHS8v-9Tx:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl.go#L47)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

The state of the closure returned by this function is immutable and bound to variables `u` and `d`. This means that prior to each call to `Request`, the callback needs to be constructed to bind the variables:

```
plder = NewPldF(u, d)
ctrl_msg.Request(ctx, plder, ...)
```

At that point, why not just:
```
pld, err = NewPld(u, d)
// Handle err
ctrl_msg.Request(ctx, pld, ...)
```

The closure callback just moves error handling from the outside to inside `Request`, but it looks like it does nothing else.

More complex implementations of `Plder` are also subject to the same limitation, because their whole state needs to be constructed prior to each call to `Request`.
</blockquote></details>

Done.

---

*[go/lib/ctrl/ctrl.go, line 47 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46UeHVrrU2xGtOr92A:-L47-J5P8w7Hlq5Ymkq_:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl.go#L47)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

I'm not a fan of this pattern. The only thing you gain is that the caller doesn't have to first create the payload, check for an error, then call `Request` and check for an error and instead can just check the error of `Request`. However,  it took me a while to understand this. 
</blockquote></details>

Done.

---

*[go/lib/ctrl/ctrl.go, line 96 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46IZHbPrnFQWeUz0SP:-L47-abwzBHdwTm2zvkE:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl.go#L96)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

The returned values here probably need a docstring.
</blockquote></details>

Done.

---

*[go/lib/ctrl/ctrl.go, line 109 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46Ic75uhSZ362SPfBN:-L47-ayP5_XFgTho6sFG:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl.go#L109)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

The returned values here probably need a docstring.
</blockquote></details>

Done.

---

*[go/lib/ctrl/signed_util.go, line 24 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46cKNxpBXBnrbM1tsA:-L47-buu2-1LofWj2FMs:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L24)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

I wonder if all of these callbacks and closures should instead be `interfaces`. I learned that you can implement an interface with purely a method so this would allow your approach, but would also allow an object-based implementation plus I think interfaces are more intuitive to understand than closures.
</blockquote></details>

Done.

---

*[go/lib/ctrl/signed_util.go, line 36 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46ZFwREQAVTWea447E:-L47-coMEFocyjaXwYov:b-cy24yv) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L36)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

I _think_ this will lead to a circular dependency. Having the trust store here means importing `go/lib/trust`. The trust store itself will make use of `go/lib/ctrl`, either directly or through the messenger, which will then lead to a circular import. I'm not sure if these utils can be in `ctrl`
</blockquote></details>

Good point, the PR with the trust store can instead put an interface into `go/lib/infra` as @scrye suggested, and just change the type in this file.

---

*[go/lib/ctrl/signed_util.go, line 38 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46d5-XNrQMYisbuVl5:-L47-kYymRQgVn_y-ads:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L38)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Suggest changing `SigChecker` to `Verifier`, so it's more intuitive for external packages to refer to `ctrl.Signer` and `ctrl.Verifier`.
</blockquote></details>

Done.

---

*[go/lib/ctrl/signed_util.go, line 39 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46VjmPZIKOB3x-cLdZ:-L47-lLUYLGHJjdmZy1v:b-271o07) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L39)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

I'm not sure I understand what the meaning behind this type is. To me it looks like it is a variant of `SigChecker` with a bound trust store, but it is used after some checks in `MkSigchecker`. Where does the job of `SigChecker` internals stop, and the job of `SigCheckerInner` start?
</blockquote></details>

I've rewritten this, hopefully it's clearer now.

---

*[go/lib/ctrl/signed_util.go, line 41 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46ScDoga6GYcq8adXe:-L47-purcdE4U6ntylcq:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L41)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

The purpose of `sigCin` is not evident to me. A very detailed docstring is needed here, s.t. users of package `ctrl` have an intuition on the differences between `SigChecker` and `SigCheckerInner`.
</blockquote></details>

Done.

---

*[go/lib/ctrl/signed_util.go, line 41 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46Y9sqGyunQWaOPPon:-L47-qEy70Kowgjf-MZp:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L41)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

`MkSigChecker`
</blockquote></details>

Done.

---

*[go/lib/ctrl/signed_util.go, line 94 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46_ct33uPO-Mklx2QI:-L47-qw7ee0zcA3Lxivg:b-9x38du) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/signed_util.go#L94)):*
<details><summary><i>Previously, shitz wrote…</i></summary><blockquote>

This is a bit too acronym-y for my tasted. `MkSignedPld`is reasonably short. Also, given how often I see `ctrl.MkSPld(..., ctrl.NullSigner)` it probably would make sense to have something like `MkUnsignedPld`, even though it might be confusing that it returns an object of `SignedPld`. I guess the correct term would be `MkUnsignedSignedPld`, and the correct name for `SignedPld` would be `PossiblySignedPld` ;)
</blockquote></details>

I realised that if Plder is no longer an interface, there's no need for this to be a function, and so i just added a `SignedPld` method to `Pld` instead that takes a Signer`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1385)*
<!-- Sent from Reviewable.io -->

shitz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1385#-:-L47z1Gs6DxB06uIpaiA:bnfp4nl)*
<!-- Sent from Reviewable.io -->

kormat commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[go/lib/ctrl/ctrl_msg/msg.go, line 15 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46eDH0041pSOqiXHPq:-L4AbFsx5ZLjci6rrBC0:b-896fix) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl_msg/msg.go#L15)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

This needs a docstring, as it's not immediately clear what the package offers (it's more related to signing and verifying, than actually messages).
</blockquote></details>

Done.

---

*[go/lib/ctrl/ctrl_msg/msg.go, line 78 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46ffeeiplQEa1sBu1I:-L4AbK7G7-O5zxvpPxWb:bjllu3h) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl_msg/msg.go#L78)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

I think the callback pattern is a bit more confusing than just calling a helper method to set up `spld` at the start of both `Notify` and `NotifyUnreliable`, and then calling the correct function. The reason being that to understand`Notify` and `NotifyUnreliable` a programmer needs to look into `notify`, and the other way around. In any case, I guess it's about preference here, and I prefer only using callbacks only when there are no other solutions.
</blockquote></details>

Given that the entire 3 methods are 15 lines in total, i don't really see much room for confusion in this case.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1385)*
<!-- Sent from Reviewable.io -->

scrye commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[go/lib/ctrl/ctrl_msg/msg.go, line 78 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46ffeeiplQEa1sBu1I:-L4AcBbDt-FSrt7vGpej:bb26lb4) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl_msg/msg.go#L78)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

Given that the entire 3 methods are 15 lines in total, i don't really see much room for confusion in this case.
</blockquote></details>

Exactly, using function pointers (and additional types) for something so simple feels overkill. It's a matter of preference/coding style, so it doesn't need to block the merge.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1385#-:-L4Aep6AyK6ZuJ_xRW26:bnfp4nl)*
<!-- Sent from Reviewable.io -->

scrye commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[go/lib/ctrl/ctrl_msg/msg.go, line 78 at r1](https://reviewable.io:443/reviews/scionproto/scion/1385#-L46ffeeiplQEa1sBu1I:-L4AfBKnbhZtmBK06WRo:bm2ydin) ([raw file](https://github.com/scionproto/scion/blob/68b91a74ea7b71ac4145767ad77f2b3ebfe500da/go/lib/ctrl/ctrl_msg/msg.go#L78)):*
<details><summary><i>Previously, scrye (Sergiu Costea) wrote…</i></summary><blockquote>

Exactly, using function pointers (and additional types) for something so simple feels overkill. It's a matter of preference/coding style, so it doesn't need to block the merge.
</blockquote></details>

FYI Switch test

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1385)*
<!-- Sent from Reviewable.io -->

scrye commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/scionproto/scion/1385#-:-L4Akgn-jbtrhdhhZknG:bnfp4nl)*
<!-- Sent from Reviewable.io -->
